### PR TITLE
feat(expr): support variant type

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Value of variant type is stored in `bson` byte slice. When evaluating the variant value, it should be parsed just in time.

We're going to move to jsonb encode eventually, but since it's unimplemented currently, bson is used as a placeholder. cc @b41sh 

Related to https://github.com/datafuselabs/databend/pull/7522
Fixes https://github.com/datafuselabs/databend/issues/6634

### CODE REVIEW

File: src/query/functions-v2/tests/it/scalars/testdata/variant.txt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+optimized expr : NULL
+output type    : NULL
+output domain  : {NULL}
+output         : NULL
+
+
+error: 
+  --> SQL:1:1
+  |
+1 | parse_json('nuLL')
```
Comment (andylokandy) [inline src/query/functions-v2/tests/it/scalars/testdata/variant.txt:13]: @sundy-li Is this expected?

File: src/query/functions-v2/tests/it/scalars/testdata/variant.txt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  | ^^^^^^^^^^^^^^^^^^ unable to parse 'nuLL': expected ident at line 1 column 3
```
Comment (andylokandy) [inline src/query/functions-v2/tests/it/scalars/testdata/variant.txt:14]: @sundy-li Is this expected?

Comment (sundy-li) [reply src/query/functions-v2/tests/it/scalars/testdata/variant.txt:14]: Yes

File: src/query/expression/src/types/variant.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+/// BSON bytes representation of `{v: null}`.
+pub const DEFAULT_BSON: &[u8] = &[0x08, 0x00, 0x00, 0x00, 0x0A, 0x76, 0x00, 0x00];
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct VariantType;
+
+impl ValueType for VariantType {
+    type Scalar = Vec<u8>;
+    type ScalarRef<'a> = &'a [u8];
```
Comment (b41sh) [inline src/query/expression/src/types/variant.rs:37]: Can we use `RawDocumentBuf` and `RawDocument` here? They are wrappers of `Vec<u8>` and `[u8]`.
```rust
pub struct RawDocumentBuf {
    data: Vec<u8>,
}
pub struct RawDocument {
    data: [u8],
}
```

**Review by leiysky — APPROVED** [review]

File: src/query/expression/src/types/variant.rs (hunk 3 above)
Comment (andylokandy) [reply src/query/expression/src/types/variant.rs:37]: There are two reasons `RawDocument` is not used for `Scalar` here:

1.  I reused `StringColumn` for variant columns, it becomes difficult to wrap the result from `StringColumn` into `RawDocument`.
2. 2, the scalar function can easily wrap inputs into `RawDocument` on the fly.

Comment (b41sh) [reply src/query/expression/src/types/variant.rs:37]: ok

**Review by b41sh — APPROVED** [review]

**Pull request merged**